### PR TITLE
feat: implement symphony provider, perps provider factory, perps data processor, and monitoring service (+ tests)

### PR DESCRIPTION
**Pull request by mzkrasner**
# Perpetual Futures Services Implementation

## Overview
This PR implements the core services layer for perpetual futures (perps) support, building on the previously established database schema and repository layer. The implementation follows established patterns for dependency injection, error handling, and batch processing while maintaining high code quality standards with comprehensive test coverage.

### 1. Symphony Perps Provider (`apps/api/src/services/providers/perps/symphony-perps.provider.ts`)
Implemented the `IPerpsDataProvider` interface for Symphony API integration:

**Features:**
- Fetches account summaries with all trading metrics (PnL, volume, fees, ROI)
- Retrieves open positions with comprehensive details
- Supports transfer history for self-funding detection
- Implements health check endpoint for provider status monitoring
- Rate limiting (100ms between requests) to respect API limits
- Retry logic with exponential backoff (3 retries, no retries on 4xx errors)
- Robust error handling with detailed logging
- Data validation with nullish coalescing for missing fields
- Date parsing with fallback to null for invalid timestamps

**Key Implementation Details:**
- Properly typed interfaces for all Symphony API responses
- Transforms Symphony data format to standardized internal format
- Handles scientific notation in numeric values
- Zero values explicitly converted to "0" instead of null
- NaN values properly handled (converted to null, then to "0" in transforms)

**Test Coverage:**
- All happy path scenarios
- Error handling (network failures, invalid responses)
- Retry mechanism validation
- Data transformation accuracy
- Health check functionality

### 2. Perps Provider Factory (`apps/api/src/services/providers/perps-provider.factory.ts`)
Factory pattern implementation for creating perps data providers:

**Features:**
- Supports multiple provider types (external_api, onchain_indexing, hybrid)
- Currently implements Symphony provider with extensibility for future providers
- Provider health validation with optional health check support
- Comprehensive error handling for unsupported providers
- Detailed logging for provider creation and validation

**Test Coverage:**
- Symphony provider creation
- Health check validation
- Error handling for unsupported provider types
- Mock provider support for testing

### 3. Perps Data Processor Service (`apps/api/src/services/perps-data-processor.service.ts`)
Orchestrates perps data fetching, transformation, and storage:

**Features:**
- Processes individual agents and batch operations
- Competition-wide agent processing with full orchestration
- Automatic portfolio snapshot creation with NaN protection
- Efficient batch processing with controlled concurrency (10 agents per batch)
- Single-pass data structure building for optimal performance
- Promise.allSettled for resilient batch processing
- Integrated monitoring support when self-funding threshold is configured

**Performance Optimizations:**
- Batch size control for external API calls
- Parallel processing where possible
- Efficient data structure usage (Maps for O(1) lookups)
- Single iteration for building multiple data structures
- Batch database operations with chunking for large arrays

**Data Handling:**
- Robust number-to-string conversion handling scientific notation and NaN
- Explicit zero value handling ("0" instead of null)
- Comprehensive data transformation with type safety
- Raw data preservation for monitoring reuse
- Portfolio snapshot validation to prevent NaN values

**Error Handling Patterns:**
- Promise.all for tightly coupled operations (competition validation)
- Promise.allSettled for independent operations (agent processing)
- Clear documentation of architectural decisions via inline comments

**Test Coverage:**
- Individual and batch processing
- Competition orchestration (processPerpsCompetition)
- Error handling and partial failures
- Edge cases (scientific notation, zero values, NaN, missing fields)
- Provider validation
- Threshold validation scenarios

### 4. Perps Monitoring Service (`apps/api/src/services/perps-monitoring.service.ts`)
Detects self-funding violations in perpetual futures competitions:

**Detection Methods:**
1. **Transfer History Analysis**
   - Detects individual deposits exceeding threshold
   - Catches "structuring" (multiple small deposits totaling above threshold)
   - Smart confidence levels based on deposit patterns
   - Default threshold of $0 (no deposits allowed during competition)

2. **Balance Reconciliation**
   - Compares expected vs actual equity
   - Detects unexplained balance increases
   - Handles negative PnL scenarios
   - Configurable thresholds with sensible defaults

**Key Features:**
- Batch processing for multiple agents with efficient alert fetching
- N+1 query prevention through batch database operations
- Resilient error handling (one agent failure doesn't affect others)
- Detailed alert creation with evidence preservation
- Existing alert checking to avoid duplicates
- Comprehensive logging for audit trails
- Single interface for monitoring (monitorAgentsWithData)

**Performance Characteristics:**
- O(n) time complexity
- Parallel processing with Promise.allSettled
- Batch alert fetching and creation
- Map-based O(1) lookups
- Optimized database calls (batch operations with chunking)

**Test Coverage:**
- 47 comprehensive test cases covering:
  - Transfer history detection
  - Balance reconciliation
  - Structuring detection
  - Edge cases and error scenarios
  - Batch processing
  - Configuration handling
  - Pre-fetched vs fresh data scenarios

### 5. Repository Enhancements (`apps/api/src/database/repositories/perps-repository.ts`)
- Batch fetching of self-funding alerts to prevent N+1 queries
- Efficient chunking for large agent arrays (prevents SQL query limits)
- Comprehensive transaction support for data consistency

### Type Safety
- No use of TypeScript `any` types in production code
- Comprehensive type definitions for all data structures
- Proper type narrowing and validation throughout
- Single source of truth for alert types (no overlapping definitions)
- Defensive programming with nullish coalescing operators

### Code Organization
- Clear separation of concerns between services
- PerpsDataProcessor orchestrates data flow and monitoring
- PerpsMonitoringService focuses on violation detection
- Providers handle external API interactions
- Repository layer manages all database operations

### Error Handling
- Graceful degradation (e.g., if transfer history fails, balance reconciliation continues)
- Detailed error reporting with context
- Partial success support with detailed results
- Clear architectural decisions documented inline

### Code Quality
- Follows established repository patterns
- Comprehensive TSDoc documentation (99% coverage target)
- Consistent error handling and logging
- Efficient batch operations throughout
- Defensive programming with validation
- No redundant operations or unnecessary data transformations

## Testing Summary
All services have comprehensive unit test coverage:
- **Symphony Provider**: 15 tests covering all methods and edge cases
- **Provider Factory**: 11 tests for factory patterns and validation
- **Data Processor**: 31 tests including edge cases and NaN handling
- **Data Processor Orchestration**: 15 tests for processPerpsCompetition
- **Monitoring Service**: 47 tests covering all detection methods and scenarios

Total: 119 tests, all passing

## Performance Impact
- Efficient batch processing reduces database load
- N+1 query prevention through batch operations
- Single-pass data transformations where possible
- Optimized Map usage for O(1) lookups
- Chunking prevents SQL query size limits

## Next Steps
With the services layer complete, the remaining work includes:
1. Create sync-perps-data.ts script with cron scheduling
2. Modify CompetitionManager.startCompetition() to check competition type
3. Modify take-portfolio-snapshots.ts cron job to check competition type
4. Create API endpoints for perps data in competition controller
5. Update frontend components for perps competitions

### CODE REVIEW

File: apps/api/src/services/providers/perps/symphony-perps.provider.ts [hunk 1]
```diff
@@ -0,0 +1,504 @@
...
+      // Validate and provide defaults for missing required fields
+      const as = data.accountSummary;
+
+      // Log warning if critical fields are missing
+      const criticalFields = ["totalEquity", "initialCapital", "totalPnl"];
+      const missingFields = criticalFields.filter(
+        (field) =>
+          as[field as keyof typeof as] === undefined ||
+          as[field as keyof typeof as] === null,
+      );
```
[the remark this answers is not in the document]
Comment (mzkrasner) [reply apps/api/src/services/providers/perps/symphony-perps.provider.ts]: Good - fixed in [f4d82a2](https://github.com/recallnet/js-recall/pull/1198/commits/f4d82a27c2bd7f32a8d568b724cf3b14d642204d)

File: packages/db-schema/src/trading/defs.ts [hunk 2]
```diff
@@ -370,32 +370,34 @@ export const perpsAccountSummaries = tradingComps.table(
     competitionId: uuid("competition_id").notNull(),
 
     // Account metrics from provider
+    totalEquity: numeric("total_equity").notNull(),
+    initialCapital: numeric("initial_capital"),
```
[the remark this answers is not in the document]
Comment (mzkrasner) [reply packages/db-schema/src/trading/defs.ts]: Fixed in [93f50b9](https://github.com/recallnet/js-recall/pull/1198/commits/93f50b99991a6eca4769631335d070f182884d0c)

File: apps/api/src/services/perps-monitoring.service.ts [hunk 3]
```diff
@@ -0,0 +1,505 @@
...
+ * Service for monitoring perpetual futures competitions for self-funding violations
+ * Implements both transfer history and balance reconciliation detection methods
+ */
+export class PerpsMonitoringService {
+  // Default thresholds (in USD)
+  // Competition rules: NO external deposits allowed during competition
+  private static readonly DEFAULT_TRANSFER_THRESHOLD = 0; // Any deposit > $0 is suspicious
+  private static readonly DEFAULT_RECONCILIATION_THRESHOLD = 100; // Unexplained > $100
+  private static readonly DEFAULT_CRITICAL_THRESHOLD = 500; // Critical if > $500
+
```
[the remark this answers is not in the document]
Comment (mzkrasner) [reply apps/api/src/services/perps-monitoring.service.ts:62]: Ignoring for now

File: packages/db-schema/src/trading/defs.ts [hunk 4]
```diff
@@ -369,33 +369,39 @@ export const perpsAccountSummaries = tradingComps.table(
...
+    roiPercent: numeric("roi_percent"),
 
     // Account status
-    accountStatus: varchar("account_status", { length: 20 })
-      .notNull()
-      .default("active"),
+    accountStatus: varchar("account_status", { length: 20 }),
+
+    // Store raw provider response for audit
+    rawData: jsonb("raw_data"),
```
[the remark this answers is not in the document]
Comment (mzkrasner) [reply packages/db-schema/src/trading/defs.ts:405]: It does not create a new migration file

File: apps/api/src/services/perps-data-processor.service.ts [hunk 5]
```diff
@@ -0,0 +1,730 @@
...
+   * Convert number to string, handling scientific notation and zero values
+   */
+  private numberToString(value: number | undefined | null): string | null {
+    if (value === undefined || value === null) {
+      return null;
+    }
+    // Handle NaN explicitly - treat as null
+    if (isNaN(value)) {
+      return null;
+    }
```
[the remark this answers is not in the document]
Comment (mzkrasner) [reply apps/api/src/services/perps-data-processor.service.ts:111]: Ignore - not important now

File: apps/api/src/services/perps-monitoring.service.ts [hunk 6]
```diff
@@ -0,0 +1,511 @@
...
+    } catch (error) {
+      serviceLogger.error(
+        `[PerpsMonitoringService] Failed to batch fetch alerts:`,
+        error,
+      );
+
+      // DECISION: Return empty alerts map on database error rather than failing.
+      // This allows monitoring to continue and potentially create new alerts.
+      // The risk is we might create duplicate alerts, but that's better than
+      // missing new violations due to a database query failure.
```
[the remark this answers is not in the document]
Comment (mzkrasner) [reply apps/api/src/services/perps-monitoring.service.ts:227]: Ignore - this would add significant complexity. Current approach is a valid architectural decision

File: apps/api/src/services/providers/perps/symphony-perps.provider.ts [hunk 7]
```diff
@@ -0,0 +1,514 @@
...
+
+      return transfers;
+    } catch (error) {
+      serviceLogger.error(
+        `[SymphonyProvider] Error fetching transfers for ${walletAddress}:`,
+        error,
+      );
+      // Don't throw - transfer history is optional for self-funding detection
+      return [];
```
[the remark this answers is not in the document]
Comment (mzkrasner) [reply apps/api/src/services/providers/perps/symphony-perps.provider.ts:359]: Transfer history is the preferred approach but fallback to bal reconciliation - ignore

File: apps/api/src/services/providers/perps/symphony-perps.provider.ts [hunk 8]
```diff
@@ -0,0 +1,514 @@
...
+      // Validate and provide defaults for missing required fields
+      const as = data.accountSummary;
+
+      // Helper to check for missing fields in a type-safe way
+      const isFieldMissing = (
+        obj: Record<string, unknown>,
+        field: string,
+      ): boolean => {
+        return obj[field] === undefined || obj[field] === null;
+      };
```
[the remark this answers is not in the document]
Comment (mzkrasner) [reply apps/api/src/services/providers/perps/symphony-perps.provider.ts:206]: Ignore

File: apps/api/src/services/perps-data-processor.service.ts [hunk 9]
```diff
@@ -0,0 +1,730 @@
...
+    return competition.type === "perpetual_futures";
+  }
+
+  /**
+   * High-level orchestration method for perps competitions
+   * Handles data sync and optional self-funding monitoring
+   * @param competitionId Competition ID
+   * @returns Combined results from sync and monitoring
+   */
+  async processPerpsCompetition(competitionId: string): Promise<{
```
Comment (mzkrasner) [inline apps/api/src/services/perps-data-processor.service.ts:557]: This is the most important method - the one that's integrated into the chron script in https://github.com/recallnet/js-recall/pull/1219 , responsible for 
- Getting positions and summaries from the perps data provider (in this case, Symphony)
- Running self-funding monitoring alerts
- Persisting the data

File: apps/api/src/services/providers/perps/symphony-perps.provider.ts [hunk 10]
```diff
@@ -0,0 +1,514 @@
+import axios, { AxiosInstance } from "axios";
```
Comment (mzkrasner) [inline apps/api/src/services/providers/perps/symphony-perps.provider.ts:1]: Restructured the `providers` directory to separate perps as its own category, and `price` providers as their own (hence the import adjustments on files that use those providers

File: apps/api/src/services/index.ts [hunk 11]
```diff
@@ -37,6 +38,7 @@ class ServiceRegistry {
   private _emailService: EmailService;
   private _tradingConstraintsService: TradingConstraintsService;
   private _competitionRewardService: CompetitionRewardService;
+  private _perpsDataProcessor: PerpsDataProcessor;
```
Comment (mzkrasner) [inline apps/api/src/services/index.ts:41]: New monitoring service intentionally not instantiated here as its methods are imported and used by perpsDataProcessor

File: apps/api/src/services/perps-monitoring.service.ts [hunk 12]
```diff
@@ -0,0 +1,511 @@
...
+          t.to.toLowerCase() === walletAddress.toLowerCase() &&
+          t.timestamp > competitionStartDate,
+      );
+
+      if (allDeposits.length === 0) {
+        return null;
+      }
+
+      // Calculate total deposited
+      const totalDeposited = allDeposits.reduce((sum, t) => sum + t.amount, 0);
```
Comment (smrz2001) [inline apps/api/src/services/perps-monitoring.service.ts:344]: JS floating-point arithmetic can have precision issues.

It doesn't look like it matters if we're exactly accurate here, though. WDYT?

File: apps/api/src/services/providers/perps/symphony-perps.provider.ts [hunk 13]
```diff
@@ -0,0 +1,514 @@
...
+  /**
+   * Get account summary - transforms Symphony response to generic format
+   */
+  async getAccountSummary(walletAddress: string): Promise<PerpsAccountSummary> {
+    const startTime = Date.now();
+
+    try {
+      await this.enforceRateLimit();
+
+      serviceLogger.debug(
```
Comment (smrz2001) [inline apps/api/src/services/providers/perps/symphony-perps.provider.ts:191]: Keep logging full wallet addresses for debugging/traceability, or partially mask for privacy? Happens in a few places.

WDYT?

File: apps/api/src/services/providers/perps/symphony-perps.provider.ts [hunk 14]
```diff
@@ -0,0 +1,514 @@
...
+
+        // Performance metrics - these are optional
+        roi: as.performance?.roi,
+        roiPercent: as.performance?.roiPercent,
+        averageTradeSize: as.performance?.averageTradeSize,
+
+        // Status - use 'unknown' as default
+        accountStatus: as.accountStatus ?? "unknown",
+
+        // Store raw data for debugging
```
Comment (smrz2001) [inline apps/api/src/services/providers/perps/symphony-perps.provider.ts:254]: Probably ok, just wanted to double-check that we're cool to keep storing raw data through a burst of usage, or if we want to control via env var? Tradeoff with controlling via env var is that we'll only be able to turn it on *after* a problem has already happened.


This would also be a great use of Sentry breadcrumbs (used for debugging):
```typescript
 Sentry.addBreadcrumb({
  category: 'symphony.api',
  message: 'Account summary',
  level: 'info',
  data: {
    walletAddress: walletAddress,
    .
    .
});
```
Or via Sentry sampling:
```typescript
const SAMPLING_RATE = 0.01; // 1% of requests

if (Math.random() < SAMPLING_RATE) {
  Sentry.captureMessage('Symphony API Response Sample', {
    level: 'debug',
    extra: {
      response: data,
      walletAddress: walletAddress,
      .
      .
    }
  });
}
```

File: apps/api/src/services/perps-data-processor.service.ts [hunk 15]
```diff
@@ -0,0 +1,730 @@
...
+    // Handle NaN explicitly - treat as null
+    if (isNaN(value)) {
+      return null;
+    }
+    // Handle zero explicitly
+    if (value === 0) {
+      return "0";
+    }
+
+    // For very small or very large numbers, use a different approach
```
Comment (smrz2001) [inline apps/api/src/services/perps-data-processor.service.ts:117]: Would strongly recommend using `decimal.js` here and in a couple of other places we're making floating-point operations. For example:
```typescript
import { Decimal } from 'decimal.js';

Decimal.set({
  precision: 40,
  rounding: Decimal.ROUND_HALF_UP,
  minE: -40,
  maxE: 40,
  toExpNeg: -40,
  toExpPos: 40
});
```

File: apps/api/src/services/perps-monitoring.service.ts (hunk 12 above)
Comment (smrz2001) [reply apps/api/src/services/perps-monitoring.service.ts:344]: Perhaps another opportunity to use `decimal.js` for accuracy.

File: apps/api/src/services/providers/perps/symphony-perps.provider.ts (hunk 14 above)
Comment (mzkrasner) [reply apps/api/src/services/providers/perps/symphony-perps.provider.ts:254]: I might've went a little overkill, but let me know if this is what you had in mind: [070e74d](https://github.com/recallnet/js-recall/pull/1225/commits/070e74df2d8504bed56a0d9da240fe57daafd9b2)

File: apps/api/src/services/perps-monitoring.service.ts (hunk 12 above)
Comment (mzkrasner) [reply apps/api/src/services/perps-monitoring.service.ts:344]: Much agreed - implemented in the last (final) pr: [c49c299](https://github.com/recallnet/js-recall/pull/1225/commits/c49c299630b3cf07624f137a7844d6c351af0d83)

File: apps/api/src/services/providers/perps/symphony-perps.provider.ts (hunk 13 above)
Comment (mzkrasner) [reply apps/api/src/services/providers/perps/symphony-perps.provider.ts:191]: ah agreed - also implemented in final pr: [a7fb8f3](https://github.com/recallnet/js-recall/pull/1225/commits/a7fb8f306254fe33710949a6594dd41de1515015)

File: apps/api/src/services/perps-data-processor.service.ts (hunk 15 above)
Comment (mzkrasner) [reply apps/api/src/services/perps-data-processor.service.ts:117]: Great catch - implemented in the last pr: [c49c299](https://github.com/recallnet/js-recall/pull/1225/commits/c49c299630b3cf07624f137a7844d6c351af0d83)

File: apps/api/src/services/perps-data-processor.service.ts [hunk 16]
```diff
@@ -0,0 +1,730 @@
...
+      let competitionStartDate: Date | null = null;
+      if (this.shouldRunMonitoring(earlyThreshold)) {
+        if (!competition.startDate) {
+          throw new Error(
+            `Competition ${competitionId} has no start date, cannot process perps data`,
+          );
+        }
+        competitionStartDate = competition.startDate; // Now safely non-null
+
+        if (competitionStartDate > new Date()) {
```
Comment (smrz2001) [inline apps/api/src/services/perps-data-processor.service.ts:609]: Should we continue from here if the competition hasn't started yet?

File: apps/api/src/services/providers/perps/symphony-perps.provider.ts (hunk 14 above)
Comment (smrz2001) [reply apps/api/src/services/providers/perps/symphony-perps.provider.ts:254]: That works! We can reduce Sentry usage if it starts adding up.

File: apps/api/src/services/perps-data-processor.service.ts (hunk 16 above)
Comment (mzkrasner) [reply apps/api/src/services/perps-data-processor.service.ts:609]: Agreed that we should be defensive here - added this in: [40f3fb0](https://github.com/recallnet/js-recall/pull/1225/commits/40f3fb0b87d74c024d5a08c81b4b2c6986f7d649)

Just a note that the chron script (included in the next PR) is the only process that would call out to this service and already has guards that prevent it from running if there is not an active competition, but I added this in anyway for good practice
